onnx/common/path.h

```diff
@@ -8,17 +8,67 @@
 #pragma once
 
 #include <string>
+#ifdef _WIN32
+#include <windows.h>
+#endif
 
 namespace ONNX_NAMESPACE {
 
 #ifdef _WIN32
 const std::string k_preferred_path_separator = "\\";
+const std::wstring w_k_preferred_path_separator = L"\\";
 #else // POSIX
 const std::string k_preferred_path_separator = "/";
 #endif
 
-std::string path_join(const std::string& origin, const std::string& append);
-void normalize_separator(std::string& path);
-std::string clean_relative_path(const std::string& path);
+template <typename STRING>
+STRING path_join(const STRING& origin, const STRING& append);
+
+template <typename CHAR>
+bool is_path_separator(CHAR c) {
+  // Windows accept / as path separator.
+  if (k_preferred_path_separator == "\\") {
+    return c == '\\' || c == '/';
+  }
+
+  return c == k_preferred_path_separator[0];
+}
+
+// To always use backward slash as path separator on Windows
+template <typename STRING>
+void normalize_separator(STRING& path) {
+  char preferred_sep = k_preferred_path_separator[0];
+  if (preferred_sep == '/') {
+    // Do nothing on linux.
+    return;
+  }
+
+  for (size_t i = 0; i < path.size(); i++) {
+    if (is_path_separator(path[i]) && path[i] != preferred_sep) {
+      path[i] = preferred_sep;
+    }
+  }
+}
+
+// Clean up relative path when there is ".." in the path, e.g.: a/b/../c -> a/c
+// std::string or std::wstring
+template <typename STRING>
+STRING clean_relative_path(const STRING& path);
+
+// std::string in template cannot be recognized as char[]
+// Therefore explicitly define char[] to handle char[] input
+template <typename CHAR, typename std::size_t N>
+std::basic_string<CHAR> clean_relative_path(const CHAR (&path)[N]) {
+  return clean_relative_path(std::basic_string<CHAR>(path));
+}
+
+#ifdef _WIN32
+inline std::wstring utf8str_to_wstring(const std::string& utf8str) {
+  int size_required = MultiByteToWideChar(CP_UTF8, 0, utf8str.c_str(), (int)utf8str.size(), NULL, 0);
```

Comment:
Please check the range of utf8str.size() before forcing converting it to int. size_t could be bigger than int.